example/src/main/java/org/mozilla/fxa_data/example/FirefoxDataInRecyclerViewExampleActivity.kt

```diff
@@ -0,0 +1,233 @@
+    private class LoginCallback constructor(activity: FirefoxDataInRecyclerViewExampleActivity) : FirefoxDataLoginManager.LoginCallback {
+        private val activityWeakReference: WeakReference<FirefoxDataInRecyclerViewExampleActivity> = WeakReference(activity)
+
+        override fun onSuccess(dataClient: FirefoxDataClient) {
+            val activity = activityWeakReference.get() ?: return
+
+            activity.isWaitingForCallback = false
+            val result: FirefoxDataResult<List<HistoryRecord>>
+            try {
+                result = dataClient.allHistory
+            } catch (e: FirefoxDataException) {
+                activity.updateUI(UIState.ERROR, e)
+                return
+            }
+
+            activity.historyAdapter.setHistoryRecords(result.result)
+            activity.updateUI(UIState.SHOW_DATA)
+        }
+
+        override fun onFailure(e: FirefoxDataException) {
+            val activity = activityWeakReference.get() ?: return
+
+            activity.updateUI(UIState.ERROR, e)
+        }
+
+        override fun onUserCancel() {
+            val activity = activityWeakReference.get() ?: return
+
+            activity.isWaitingForCallback = false
+            activity.updateUI(UIState.SIGN_IN_PROMPT)
+        }
+    }
+
+    val signInAction = { v: View ->
+        loginManager.promptLogin(this@FirefoxDataInRecyclerViewExampleActivity,
+                resources.getString(R.string.app_name),
+                LoginCallback(this@FirefoxDataInRecyclerViewExampleActivity))
+        isWaitingForCallback = true
+        updateUI(UIState.LOADING)
+
+    }
+
+
+    val signOutAction: (View) -> Unit = {
+        loginManager.signOut()
+        updateUI(UIState.SIGN_IN_PROMPT)
+    }
+
+
+    private fun updateUI(uiState: UIState, exception: Exception? = null) {
+        runOnUiThread {
+            // Reset UI state by removing all mutations we make.
+            for (container in containerViews) {
+                container.visibility = View.GONE // we'll unhide one of these.
+            }
+
+            explanationButton.isEnabled = true
+
+            // Mutate to the new state for the given arguments.
+            if (uiState == UIState.SHOW_DATA) {
+                // Adapter takes care of data updates.
+                fxDataContainer.visibility = View.VISIBLE
+            } else if (uiState == UIState.SIGN_IN_PROMPT || uiState == UIState.ERROR) {
+                explanationContainer.visibility = View.VISIBLE
+            } else {
+                loadingContainer.visibility = View.VISIBLE
+            }
+
+            when (uiState) {
+                UIState.SIGN_IN_PROMPT -> {
+                    val res = resources
+                    setExplanationUIResources(R.string.sign_in_explanation,
+                            R.string.sign_in_button,
+                            signInAction)
+                }
+
+                UIState.ERROR -> {
+                    val res = resources
```

Comment:
nit: `res` isn't much shorter than `resources` – I'd prefer if we just used `resources` directly